### PR TITLE
Fix errors in some floating point cases

### PR DESCRIPTION
**Pull request by barendgehrels**
* in intersection a check is made for overlap (solving some errors mainly in small scale AND nearly collinear situations)
* in touch a special case is added where intersection reports touching segments, but they are just not touching
* in touch some sides were reported as 0 while, in FP, they are distinguishable and that can be used to fix some errors

### CODE REVIEW

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 1]
```diff
@@ -111,6 +111,60 @@ struct base_turn_handler
...
+
+        typedef typename select_coordinate_type
+                <
+                    typename UniqueSubRange1::point_type,
+                    typename UniqueSubRange2::point_type
+                    >::type coordinate_type;
+
+        typedef detail::distance_measure<coordinate_type> dm_type;
+
+        dm_type const dm = get_distance_measure(range_p.at(range_index), range_p.at(range_index + 1), range_q.at(point_index));
```
Comment (awulkiew) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:142]: What are these ranges? I don't remember. Shouldn't the type of the index rather be some kind of unsigned integral type like `size_t` or ideally defined by the `UniqueSubRange` type as e.g. `size_type`?

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 2]
```diff
@@ -369,14 +423,80 @@ struct touch : public base_turn_handler
...
+        // |  \qj       In this case it should block P and
+        // |  ^------   set Q to Union
+        // >----->P     qj is LEFT of P1 and pi is LEFT of Q2
+        //              (the other way round is also possible)
+
+        typedef typename select_coordinate_type
+                <
+                    typename UniqueSubRange1::point_type,
+                    typename UniqueSubRange2::point_type
+                    >::type coordinate_type;
```
Comment (awulkiew) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:462]: Too much indentation. ;)

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 1 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:142]: We pass small ranges to be able to know the next points (point after two intersecting segments).
I will look what we normally do, assuming indeed we can pass size_t

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 2 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:462]: will fix, thanks & thanks for your review

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 1 above)
Comment (awulkiew) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:142]: So the size of the range is 1, 2 or 3. Ok, bigger type is not needed then.

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 3]
```diff
@@ -111,6 +111,60 @@ struct base_turn_handler
...
 
+
+#if ! defined(BOOST_GEOMETRY_USE_RESCALING)
+    template
+    <
+        typename UniqueSubRange1,
+        typename UniqueSubRange2
+    >
+    static inline int side_with_distance_measure(UniqueSubRange1 const& range_p,
+            UniqueSubRange2 const& range_q,
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:122]: is indentation right here?

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 4]
```diff
@@ -111,6 +111,60 @@ struct base_turn_handler
...
+        if (point_index >= 2 && range_q.is_last_segment())
+        {
+            return 0;
+        }
+
+        typedef typename select_coordinate_type
+                <
+                    typename UniqueSubRange1::point_type,
+                    typename UniqueSubRange2::point_type
+                    >::type coordinate_type;
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:138]: Too much indentation also here

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 5]
```diff
@@ -369,14 +423,80 @@ struct touch : public base_turn_handler
...
 
-    /*static inline void block_second(bool block, TurnInfo& ti)
+#if ! defined(BOOST_GEOMETRY_USE_RESCALING)
+    template
+    <
+        typename UniqueSubRange1,
+        typename UniqueSubRange2
+    >
+    static inline bool handle_imperfect_touch(UniqueSubRange1 const& range_p,
+            UniqueSubRange2 const& range_q, TurnInfo& ti)
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:433]: indentation issue?

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 1 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:142]: Indeed, OK, thanks

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 5 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:433]: Well, this is how we (at least I) did it from the beginning - an indentation of 8. However, currently IDE's usually automatically align it under the next bracket and some features (`clang_format`) or people also do this. So now we've a mixed approach.

IMO both are not perfect... For the first method it is currently unusual, I think. For the second method, with long names and modifiers, it goes really to the right and needs more lines.

I keep this for now, if you've strong opinions, just comment.

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 4 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:138]: Thanks, indeed. Fixed now, all in this file (there were some other ones too somehow)

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp (hunk 2 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp]: all fixed, amended and pushed

**Pull request merged**